esphome_nimble/components/nimble_tracker/queue.h

```diff
@@ -0,0 +1,65 @@
+// Copied from https://github.com/esphome/esphome/blob/dev/esphome/components/esp32_ble_tracker/queue.h
+#pragma once
+
+#include "esphome/core/component.h"
+#include "esphome/core/helpers.h"
+
+#include <queue>
+#include <mutex>
+#include <cstring>
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/semphr.h>
+
+/*
+ * BLE events come in from a separate Task (thread) in the ESP32 stack. Rather
+ * than trying to deal with various locking strategies, all incoming GAP and GATT
+ * events will simply be placed on a semaphore guarded queue. The next time the
+ * component runs loop(), these events are popped off the queue and handed at
+ * this safer time.
+ */
+
+namespace esphome
+{
+  namespace nimble_tracker
+  {
+
+    template <class T>
+    class Queue
+    {
+    public:
+      Queue() { m_ = xSemaphoreCreateMutex(); }
+
+      void push(T *element)
+      {
+        if (element == nullptr)
+          return;
+        if (xSemaphoreTake(m_, 5L / portTICK_PERIOD_MS))
```

Comment:
Shouldn't it be a `while(!` here instead of `if` to avoid leak and loosing element if the take failed ?

Reply:
Honestly no idea 😅, the original code is here https://github.com/esphome/esphome/blob/dev/esphome/components/esp32_ble_tracker/queue.h